daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala

```diff
@@ -4,395 +4,53 @@
-                recordedFields.length == replayedFields.length &&
-                (keys(recordedFields) zip keys(replayedFields)).forall(nameIsReplayedBy) &&
-                loop((values(recordedFields) zip values(recordedFields)) ++: rest)
-            case (
-                  ValueVariant(recordedTyCon, recordedVariant, recordedValue),
-                  ValueVariant(replayedTyCon, replayedVariant, replayedValue),
-                ) =>
-              tyConIsReplayedBy(recordedTyCon, replayedTyCon) &&
-                recordedVariant == replayedVariant &&
-                loop((recordedValue, replayedValue) +: rest)
-            case (ValueList(recordedValues), ValueList(replayedValues)) =>
-              recordedValues.length == replayedValues.length &&
-                loop((recordedValues.iterator zip replayedValues.iterator) ++: rest)
-            case (ValueOptional(recordedValue), ValueOptional(replayedValue)) =>
-              (recordedValue, replayedValue) match {
-                case (Some(recorded), Some(replayed)) => loop((recorded, replayed) +: rest)
-                case (None, None) => loop(rest)
-                case _ => false
-              }
-            case (ValueTextMap(recordedEntries), ValueTextMap(replayedEntries)) =>
-              recordedEntries.length == replayedEntries.length &&
-                (keys(recordedEntries.toImmArray) sameElements keys(replayedEntries.toImmArray)) &&
-                loop(
-                  (values(recordedEntries.toImmArray) zip values(replayedEntries.toImmArray)) ++:
-                    rest
-                )
-            case (ValueGenMap(recordedEntries), ValueGenMap(replayedEntries)) =>
-              recordedEntries.length == replayedEntries.length &&
-                loop(
-                  (keys(recordedEntries) zip keys(replayedEntries)) ++:
-                    (values(recordedEntries) zip values(replayedEntries)) ++:
-                    rest
-                )
-            case _ =>
-              false
-          }
-        case LazyList() =>
-          true
-      }
-    loop(LazyList((recorded, replayed)))
-  }
+private final class Validation[Nid, Cid]() {
 
   /** Whether `replayed` is the result of reinterpreting this transaction.
     *
     * @param recorded : the transaction to be validated.
     * @param replayed : the transaction resulting from the reinterpretation of
     *   the root nodes of [[recorded]].
-    * @note This function is asymmetric in order to provide backward compatibility.
-    *      For instance, some field may be undefined in the [[recorded]] transaction
-    *      while present in the [[replayed]] one.
+    * @note This function is symmetric
+    *
+    * 'isReplayedBy' normalizes both arguments, and then determines structural equality.
+    *
+    * The RIGHT `replayed` arg must be normalized because the engine currently does not.
+    *
+    * The LEFT `recorded` arg is normalized for the sake of test cases which do not ensure
```

Comment:
can we do this at the call site only in tests (potentially via a separate helper function) to avoid the performance impact on a real ledgers?